src/plugins/teleop/Teleop.cc

```diff
@@ -0,0 +1,274 @@
+ * limitations under the License.
+ *
+*/
+
+#include <iostream>
+#ifdef _MSC_VER
+#pragma warning(push, 0)
+#endif
+#include <ignition/msgs/twist.pb.h>
+#ifdef _MSC_VER
+#pragma warning(pop)
+#endif
+
+#include <string>
+
+#include <ignition/gui/Application.hh>
+#include <ignition/gui/MainWindow.hh>
+#include <ignition/plugin/Register.hh>
+
+#include "Teleop.hh"
+
+namespace ignition
+{
+namespace gui
+{
+namespace plugins
+{
+  enum class keyLinear{
+    forward,
+    backward,
+    stop,
+  };
+
+  enum class keyAngular{
+    left,
+    right,
+    stop,
+  };
+
+  class TeleopPrivate
+  {
+    /// \brief Node for communication.
+    public: ignition::transport::Node node;
+
+    /// \brief Topic. Set '/cmd_vel' as default.
+    public: std::string topic = "/cmd_vel";
+
+    /// \brief Publisher.
+    ignition::transport::Node::Publisher cmdVelPub;
+
+    /// \brief Linear velocity.
+    public: float linearVel = 0;
+    /// \brief Angular velocity.
+    public: float angularVel = 0;
+
+    /// \brief Linear direction.
+    public: int linearDir = 0;
+    /// \brief Angular direction.
+    public: int angularDir = 0;
+
+    /// \brief Linear state setted by keyboard input.
+    public: keyLinear linearState = keyLinear::stop;
+    /// \brief Angular state setted by keyboard input.
+    public: keyAngular angularState = keyAngular::stop;
+
+    public: bool keyEnable = false;
+    public: bool newTopic = true;
+
+  };
+}
+}
+}
+
+using namespace ignition;
+using namespace gui;
+using namespace plugins;
+
+/////////////////////////////////////////////////
+Teleop::Teleop(): Plugin(), dataPtr(new TeleopPrivate)
+{
+}
+
+/////////////////////////////////////////////////
+Teleop::~Teleop()
+{
+}
+
+/////////////////////////////////////////////////
+void Teleop::LoadConfig(const tinyxml2::XMLElement *_pluginElement)
+{
+  if (this->title.empty())
+    this->title = "Teleop";
+
+  ignition::gui::App()->findChild
+    <ignition::gui::MainWindow *>()->QuickWindow()->installEventFilter(this);
+}
+
+/////////////////////////////////////////////////
+void Teleop::OnTeleopTwist()
+{
+  ignition::msgs::Twist cmdVelMsg;
+
+  cmdVelMsg.mutable_linear()->set_x(
+      this->dataPtr->linearDir * this->dataPtr->linearVel);
+  cmdVelMsg.mutable_angular()->set_z(
+      this->dataPtr->angularDir * this->dataPtr->angularVel);
+
+  if (this->dataPtr->newTopic){
+    this->dataPtr->cmdVelPub = ignition::transport::Node::Publisher();
+    this->dataPtr->cmdVelPub =
+    this->dataPtr->node.Advertise<ignition::msgs::Twist>(this->dataPtr->topic);
+    this->dataPtr->newTopic = false;
+  }
+
+  this->dataPtr->cmdVelPub.Publish(cmdVelMsg);
+}
+
+/////////////////////////////////////////////////
+void Teleop::OnTopicSelection(const QString& _topic)
+{
+  this->dataPtr->newTopic = true;
+  this->dataPtr->topic = _topic.toStdString();
+  ignmsg << "[OnTopicSelection]: topic: " << this->dataPtr->topic << std::endl;
```

Comment:
Probably logging a message with `[OnTopicSelection]:` it isn't accurate.
Think of the messages to help the user, no the developer.

I would print something like:
```cpp
ignmsg << "New topic has been entered: ' "<< this->dataPtr->topic << " ' " << std::endl;
```

Some comment for all the other cases

Reply:
Done